Data Science Project on Train Route Analysis and Journey Time Prediction

Import Libraries

In [ ]:
import pandas as pd
import numpy as np

Load Dataset

In [ ]:
df = pd.read_csv("Dataset1.csv")

First 5 rows

In [ ]:
df.head()

,SN,Train_No,Station_Code,1A,2A,3A,SL,Station_Name,Route_Number,Arrival_time,Departure_Time,Distance
0,1,107,SWV,100,100,100,100,SAWANTWADI R,1,00:00:00,10:25:00,0
1,2,107,THVM,260,228,196,164,THIVIM,1,11:06:00,11:08:00,32
2,3,107,KRMI,345,296,247,198,KARMALI,1,11:28:00,11:30:00,49
3,4,107,MAO,490,412,334,256,MADGOAN JN.,1,12:10:00,00:00:00,78
4,1,108,MAO,100,100,100,100,MADGOAN JN.,1,00:00:00,20:30:00,0


Task 1.1 - Summary of Dataset

Total Records and Columns

In [ ]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Rows: 186074
Columns: 12


Column Names

In [ ]:
print(df.columns)

Index(['SN', 'Train_No', 'Station_Code', '1A', '2A', '3A', 'SL',
       'Station_Name', 'Route_Number', 'Arrival_time', 'Departure_Time',
       'Distance'],
      dtype='str')


Dataset Information

In [ ]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 186074 entries, 0 to 186073
Data columns (total 12 columns):
 #   Column          Non-Null Count   Dtype
---  ------          --------------   -----
 0   SN              186074 non-null  int64
 1   Train_No        186074 non-null  int64
 2   Station_Code    186074 non-null  str  
 3   1A              186074 non-null  int64
 4   2A              186074 non-null  int64
 5   3A              186074 non-null  int64
 6   SL              186074 non-null  int64
 7   Station_Name    186074 non-null  str  
 8   Route_Number    186074 non-null  int64
 9   Arrival_time    186074 non-null  str  
 10  Departure_Time  186074 non-null  str  
 11  Distance        186074 non-null  int64
dtypes: int64(8), str(4)
memory usage: 17.0 MB


Summary Statement

Dataset contains 186,074 records and 12 columns related to train routes, stations, timings and distance information.

Task 1.2: Train-wise Table Showing Route Start and End Stations

Create Route table

In [7]:
route_table = (
    df.groupby('Train_No')
      .apply(lambda x: pd.Series({
          'Start_Station': x.iloc[0]['Station_Name'],
          'End_Station': x.iloc[-1]['Station_Name'],
          'Number_of_Stops': len(x)
      }))
      .reset_index()
)

Display First 10 Records

In [8]:
route_table.head(10)

,Train_No,Start_Station,End_Station,Number_of_Stops
0,107,SAWANTWADI R,MADGOAN JN.,4
1,108,MADGOAN JN.,SAWANTWADI R,4
2,128,MADGOAN JN.,CHHATRAPATI,22
3,290,DELHI-SAFDAR,DELHI-SAFDAR,14
4,401,AURANGABAD,VARANASI JN.,12
5,421,LUCKNOW JN.,SHRI MATA VA,5
6,422,SHRI MATA VA,LUCKNOW JN.,5
7,477,SIRSA,SIRSA,14
8,502,RAJENDRANAGA,AMBALA CANTT,9
9,504,PATNA JN.,BATHINDA JN,3


Total unique trains

In [ ]:


df['Train_No'].nunique()

11113

Task 1.3: Calculate Basic Statistics for Distance and Number of Stops

Distance Statistics

In [10]:
df['Distance'].describe()

count    186074.000000
mean        281.353838
std         483.743964
min           0.000000
25%          23.000000
50%          73.000000
75%         291.000000
max        4260.000000
Name: Distance, dtype: float64

Number of Stops Statistics
  
  
  Count Stops per Train

In [11]:
stops_per_train = df.groupby('Train_No').size()

Calculate Statistics

In [12]:
stops_per_train.describe()

count    11113.000000
mean        16.743814
std         12.993123
min          2.000000
25%          8.000000
50%         15.000000
75%         22.000000
max        118.000000
dtype: float64

Task 1.4: Identify Missing, Null, or Inconsistent Values

Check Missing Values

In [13]:
df.isnull().sum()

SN                0
Train_No          0
Station_Code      0
1A                0
2A                0
3A                0
SL                0
Station_Name      0
Route_Number      0
Arrival_time      0
Departure_Time    0
Distance          0
dtype: int64

Total Missing Values

In [14]:
df.isnull().sum().sum()

np.int64(0)

Check Duplicate Records

In [15]:
df.duplicated().sum()

np.int64(0)

Check empty station Codes

In [ ]:


(df['Station_Code'] == '').sum()

np.int64(0)

Check Arrival and Departure Times


In [ ]:


df[['Arrival_time','Departure_Time']].head()

,Arrival_time,Departure_Time
0,00:00:00,10:25:00
1,11:06:00,11:08:00
2,11:28:00,11:30:00
3,12:10:00,00:00:00
4,00:00:00,20:30:00
